plugin.go

```diff
@@ -68,67 +70,62 @@ type Plugin struct {
 	update bool
 }
 
-func (p *Plugin) retrieve() (err error) {
+func (p *Plugin) updatePlugin() (err error) {
 	if len(p.gitURL) == 0 {
 		return
 	}
 
-	switch {
-	case !doesPluginSourceExist(p.gitURL):
-		p.out.Notification("Plugin source does not yet exist, downloading repository")
-		// We don't have the source yet. Download it
-		if err = goGet(p.gitURL); err != nil {
-			err = fmt.Errorf("error performing go get: %v", err)
-			return
-		}
+	if len(p.branch) > 0 {
+		if err = p.checkout(); err != nil {
+			// Check if error was expected: related to pulling a tag successfully
+			if !strings.Contains(err.Error(), "HEAD is now at") {
+				p.out.Errorf("Error checking out %s: %+v", p.branch, err)
+				return
+			}
 
-	default:
-		// We have the source already. Perform git pull to make sure the branch is synced
-		p.out.Notification("Pulling most recent version")
-		if _, err = gitPull(p.gitURL); err != nil {
-			err = fmt.Errorf("error performing git pull: %v", err)
-			return
+			p.out.Notificationf("Updated to version: %s", p.branch)
+			return p.updateDependencies()
 		}
+	}
 
-		// Update all the plugin dependencies
-		p.out.Notification("Updating plugin dependencies")
-		if err = updatePluginDependencies(p.gitURL); err != nil {
-			err = fmt.Errorf("error updating plugin dependencies: %v", err)
-			return
+	// Ensure we're up to date with the given branch
+	var status string
+	if status, err = gitPull(p.gitURL); len(status) == 0 || err != nil {
+		if err == nil {
```

Comment:
agreed. this was copied from logic in handling the tri-state "git pull successful, git pull unnecessary, git pull ignored"

this is due to the stdErr output and status code in git commands :(

Reply:
It's only checking the condition of "we didn't fail but we didn't update anything either" in order to print the notification.

we can ignore that if you prefer?

Reply:
I think we can ignore it and see if it creates any usage confusion. It might be nice to only notify if things are happening. That way we can reduce noise. What do you think?

Reply:
@itsmontoya fixed

Reply:
@itsmontoya Updated the logic so its cleaner.

If we think it's verbose, we can just remove whatever prints we feel are redundant